src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+        return uint32(quorumApkUpdates[quorumNumber].length);
+    }
+    /// @notice Returns the length of ApkUpdates for the global APK
+    function getGlobalApkHistoryLength() external view returns(uint32){
+        return uint32(globalApkUpdates.length);
+    }
+
+    function _processGlobalApkUpdate(BN254.G1Point memory point) internal {
+        // load and store in memory in common case we need to access the length again
+        uint256 globalApkUpdatesLength = globalApkUpdates.length;
+        // update the nextUpdateBlockNumber of the previous update
+        if (globalApkUpdates.length > 0) {
+            globalApkUpdates[globalApkUpdatesLength - 1].nextUpdateBlockNumber = uint32(block.number);
+        }
+
+        // accumulate the given point into the globalApk
+        globalApk = globalApk.plus(point);
+        // add this update to the list of globalApkUpdates
+        ApkUpdate memory latestGlobalApkUpdate;
+        latestGlobalApkUpdate.apkHash = BN254.hashG1Point(globalApk);
+        latestGlobalApkUpdate.updateBlockNumber = uint32(block.number);
+        globalApkUpdates.push(latestGlobalApkUpdate);
+    }
+
+    function _processQuorumApkUpdate(bytes memory quorumNumbers, BN254.G1Point memory point) internal {
+        BN254.G1Point memory apkBeforeUpdate;
+        BN254.G1Point memory apkAfterUpdate;
+
+        for (uint8 i = 0; i < uint8(quorumNumbers.length);) {
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+
+            uint256 quorumApkUpdatesLength = quorumApkUpdates[quorumNumber].length;
+            if (quorumApkUpdatesLength > 0) {
+                // update nextUpdateBlockNumber of the current latest ApkUpdate
+                quorumApkUpdates[quorumNumber][quorumApkUpdatesLength - 1].nextUpdateBlockNumber = uint32(block.number);
+            }
+            
+            apkBeforeUpdate = quorumApk[quorumNumber];
+
+            apkAfterUpdate = apkBeforeUpdate.plus(point);
+
+            //update aggregate public key for this quorum
+            quorumApk[quorumNumber] = apkAfterUpdate;
+            //create new ApkUpdate to add to the mapping
+            ApkUpdate memory latestApkUpdate;
+            latestApkUpdate.apkHash = BN254.hashG1Point(apkAfterUpdate);
+            latestApkUpdate.updateBlockNumber = uint32(block.number);
+            quorumApkUpdates[quorumNumber].push(latestApkUpdate);
+
+            unchecked{
+                ++i;
+            }
+        }
+    }
+
+    function _validateApkHashForQuorumAtBlockNumber(ApkUpdate memory apkUpdate, uint32 blockNumber) internal pure {
+        require(
+            blockNumber >= apkUpdate.updateBlockNumber, 
+            "BLSPubkeyRegistry._validateApkHashForQuorumAtBlockNumber: index too recent"
+        );
+        require(
+            apkUpdate.nextUpdateBlockNumber == 0 || blockNumber < apkUpdate.nextUpdateBlockNumber, 
+            "BLSPubkeyRegistry._validateApkHashForQuorumAtBlockNumber: not latest apk update"
+        );
```

Comment:
seems worth having an in-context comment noting that `nextUpdateBlockNumber` = 0 indicates that the update is the latest one